content/blog/20230731_upgrade-td-agent-v4-to-v5.md

```diff
@@ -0,0 +1,286 @@
+
+### 7. Install plugins added on my own.
+
+```
+$ sudo fluent-gem install fluent-plugin-concat
+```
+
+```
+$ fluent-gem list | grep fluent-plugin*
+fluent-plugin-calyptia-monitoring (0.1.3)
+fluent-plugin-concat (2.5.0)
+fluent-plugin-elasticsearch (5.3.0)
+fluent-plugin-flowcounter-simple (0.1.0)
+fluent-plugin-kafka (0.19.0)
+fluent-plugin-metrics-cmetrics (0.1.2)
+fluent-plugin-opensearch (1.1.0)
+fluent-plugin-prometheus (2.0.3)
+fluent-plugin-prometheus_pushgateway (0.1.0)
+fluent-plugin-record-modifier (2.1.1)
+fluent-plugin-rewrite-tag-filter (2.4.0)
+fluent-plugin-s3 (1.7.2)
+fluent-plugin-sd-dns (0.1.0)
+fluent-plugin-systemd (1.0.5)
+fluent-plugin-td (1.2.0)
+fluent-plugin-utmpx (0.5.0)
+fluent-plugin-webhdfs (1.5.0)
+```
+ 
+As for fluent-package v5, "fluent-plugin-concat" was installed under "/opt/fluent/lib/ruby/gems/3.2.0/gems/" directories.
+ 
+```
+$ ls -l /opt/fluent/lib/ruby/gems/3.2.0/gems/ |grep fluent-plugin*
+drwxr-xr-x.  5 root root  175  7月 14 03:14 fluent-plugin-calyptia-monitoring-0.1.3
+drwxr-xr-x.  4 root root  206  7月 14 03:16 fluent-plugin-concat-2.5.0
+drwxr-xr-x.  5 root root 4096  7月 14 03:14 fluent-plugin-elasticsearch-5.3.0
+drwxr-xr-x.  4 root root  205  7月 14 03:14 fluent-plugin-flowcounter-simple-0.1.0
+drwxr-xr-x.  6 root root  191  7月 14 03:14 fluent-plugin-kafka-0.19.0
+drwxr-xr-x.  5 root root  190  7月 14 03:14 fluent-plugin-metrics-cmetrics-0.1.2
+drwxr-xr-x.  5 root root 4096  7月 14 03:14 fluent-plugin-opensearch-1.1.0
+drwxr-xr-x.  5 root root  215  7月 14 03:14 fluent-plugin-prometheus-2.0.3
+drwxr-xr-x.  6 root root  238  7月 14 03:14 fluent-plugin-prometheus_pushgateway-0.1.0
+drwxr-xr-x.  4 root root  176  7月 14 03:14 fluent-plugin-record-modifier-2.1.1
+drwxr-xr-x.  3 root root  210  7月 14 03:14 fluent-plugin-rewrite-tag-filter-2.4.0
+drwxr-xr-x.  5 root root  230  7月 14 03:14 fluent-plugin-s3-1.7.2
+drwxr-xr-x.  3 root root  170  7月 14 03:14 fluent-plugin-sd-dns-0.1.0
+drwxr-xr-x.  3 root root   49  7月 14 03:14 fluent-plugin-systemd-1.0.5
+drwxr-xr-x.  5 root root  221  7月 14 03:14 fluent-plugin-td-1.2.0
+drwxr-xr-x.  5 root root  166  7月 14 03:14 fluent-plugin-utmpx-0.5.0
+drwxr-xr-x.  4 root root  191  7月 14 03:14 fluent-plugin-webhdfs-1.5.0
+```
+
+### 8. Start fluent-package v5 daemon.
+
+```
+$ sudo systemctl start fluent-package
+```
+
+### 9. Check if there are no error messages in fluentd logs.
+
+```
+$ tail -100f /var/log/fluent/fluentd.log
+```
+
+Now, upgrading steps are completed. Happy Logging!
+
+## Additional hints for v4 users
+
+### For Debian/Ubuntu
+
+* If you upgrade from v4, `td-agent` will be installed as a transitional package.
+  you can remove it safely with `sudo apt purge td-agent`.
+* `fluentd-apt-source` package will be marked as a transitional package.
+  you can remove it safely with `sudo apt purge fluentd-apt-source`.
```

Comment:
It looks like this says that users who upgraded from v4 need to uninstall v4.
Is this intentional?

Or do you want to say that users can uninstall v5 safely even if it is upgraded from v4?
Then, is `sudo apt purge fluent-package` correct?

Reply:
No need to uninstall in practically (it is ok to keep it), but transitional package (td-agent) is usually empty, so added note for it.

Reply:
Hmm, I think the existing v4 package is automatically uninstalled when upgrading.

I have confirmed the behavoir on RockyLinux 8.
I have upgraded from v4 to v5.
Then, I couldn't find `td-agent`.

```console
[vagrant@localhost ~]$ dnf list --installed | grep flu
Failed to set locale, defaulting to C.UTF-8
fluent-package.x86_64               5.0.0-1.el8                            @@commandline
[vagrant@localhost ~]$ dnf list --installed | grep td
Failed to set locale, defaulting to C.UTF-8
libstdc++.x86_64                    8.5.0-18.el8                           @anaconda    
libtdb.x86_64                       1.4.7-1.el8                            @anaconda    
libzstd.x86_64                      1.4.4-1.el8                            @anaconda
```

When I uninstalled `td-agent`, `fluent-package` is uninstalled.
(I didn't expect this... Looks like `td-agent` is an alias for `fluent-package`...)

```console
[vagrant@localhost ~]$ sudo dnf remove td-agent
Failed to set locale, defaulting to C.UTF-8
Dependencies resolved.
==============================================================================================================================================================================================================
 Package                                             Architecture                                Version                                             Repository                                          Size
==============================================================================================================================================================================================================
Removing:
 fluent-package                                      x86_64                                      5.0.0-1.el8                                         @@commandline                                       64 M

Transaction Summary
==============================================================================================================================================================================================================
Remove  1 Package

Freed space: 64 M
Is this ok [y/N]: y
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                                      1/1 
  Running scriptlet: fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
  Erasing          : fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
warning: /etc/fluent/fluentd.conf saved as /etc/fluent/fluentd.conf.rpmsave

  Running scriptlet: fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
Removing fluentd user...
userdel: group fluentd is the primary group of another user and is not removed.
Removing fluentd group...
groupdel: cannot remove the primary group of user 'td-agent'
Removing td-agent user...
userdel: td-agent mail spool (/var/spool/mail/td-agent) not found
userdel: td-agent home directory (/var/lib/td-agent) not found

  Verifying        : fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 

Removed:
  fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                                           

Complete!
[vagrant@localhost ~]$ dnf list --installed | grep flu
Failed to set locale, defaulting to C.UTF-8
[vagrant@localhost ~]$ fluentd --version
-bash: fluentd: command not found
[vagrant@localhost ~]$ systemctl status fluentd
Unit fluentd.service could not be found.
[vagrant@localhost ~]$ systemctl status td-agent
Unit td-agent.service could not be found.
```

Reply:
Sorry, maybe I should confirm this on Ubuntu/Debian.
I'm confirming it now.

Reply:
I have confirmed this on Ubuntu Focal.
As this document says, there seems to be the `td-agent` package remaining, but I couldn't purge it.
It seems that we need to uninstall v5 first to uninstall the remaining `td-agent` package.

```console
vagrant@vagrant:~$ apt list --installed | grep fluen

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

fluent-package/now 5.0.0-1 amd64 [installed,local]
fluentd-apt-source/focal,now 2020.8.25-1 all [installed]
vagrant@vagrant:~$ apt list --installed | grep td-agent

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

vagrant@vagrant:~$ sudo apt purge td-agent
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  td-agent*
0 upgraded, 0 newly installed, 1 to remove and 194 not upgraded.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
(Reading database ... 50480 files and directories currently installed.)
Purging configuration files for td-agent (4.5.0-1) ...
td-agent:x:113:117::/var/lib/td-agent:/usr/sbin/nologin
userdel: user td-agent is currently used by process 21510
dpkg: error processing package td-agent (--purge):
 installed td-agent package post-removal script subprocess returned error exit status 8
Errors were encountered while processing:
 td-agent
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Reply:
Hmm, it is not unexpected...

Reply:
Created as an issue to track.

https://github.com/fluent/fluent-package-builder/issues/564

Reply:
non- fluent-package issue.

Reply:
There are still concerns: https://github.com/fluent/fluent-package-builder/issues/564#issuecomment-1669220651

I suppose we can not purge v4 on the environment where the package is upgraded to v5 from v4.

Reply:
Thanks, rebased and  dropped guide to purge transitional package.